pkg/gadgets/seccomp/advisor.go

```diff
@@ -0,0 +1,236 @@
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/seccomp/types"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	libseccomp "github.com/seccomp/libseccomp-golang"
+)
+
+type SeccompAdvisor struct {
+	nodeName    string
+	bpfModule   *bpf.Module
+	syscallsMap *bpf.BPFMap
+
+	started bool
+	ticker  *time.Ticker
+	done    chan bool
+
+	// containers by mntns
+	containers        map[uint64]*types.Container
+	stoppedContainers []*types.Container
+}
+
+/* Keep in sync with constants in pkg/gadgets/seccomp/tracepoint-bpf-asset.c
+ */
+const (
+	SyscallsCount              = 500
+	SyscallsMapValueFooterSize = 1
+	SyscallsMapValueSize       = SyscallsCount + SyscallsMapValueFooterSize
+)
+
+func NewAdvisor(nodeName string) (*SeccompAdvisor, error) {
+	return &SeccompAdvisor{
+		nodeName:   nodeName,
+		containers: make(map[uint64]*types.Container),
+		done:       make(chan bool),
+	}, nil
+}
+
+func (sa *SeccompAdvisor) Start() error {
+	buf, err := Asset("tracepoint-bpf-asset.o")
+	if err != nil {
+		return fmt.Errorf("couldn't find asset: %s", err)
+	}
+
+	bpfModule, err := bpf.NewModuleFromBuffer(buf, "tracepoint-bpf-asset.o")
+	if err != nil {
+		return err
+	}
+
+	err = bpfModule.BPFLoadObject()
+	if err != nil {
+		return err
+	}
+
+	syscallsMap, _ := bpfModule.GetMap("syscalls_per_mntns")
+	blank := make([]byte, SyscallsMapValueSize)
+	err = syscallsMap.Update(uint64(0), blank)
+	if err != nil {
+		return err
+	}
+
+	prog, err := bpfModule.GetProgram("tracepoint__raw_syscalls__sys_enter")
+	if err != nil {
+		return err
+	}
+	_, err = prog.AttachRawTracepoint("sys_enter")
+	if err != nil {
+		return err
+	}
+
+	sa.bpfModule = bpfModule
+	sa.syscallsMap = syscallsMap
+
+	sa.started = true
+
+	sa.ticker = time.NewTicker(1 * time.Second)
+	go func() {
+		for {
+			select {
+			case <-sa.done:
+				return
+			case <-sa.ticker.C:
+				sa.update()
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (sa *SeccompAdvisor) Stop() {
+	if !sa.started {
+		return
+	}
+
+	sa.ticker.Stop()
+	sa.done <- true
+
+	sa.stoppedContainers = nil
+	sa.bpfModule.Close()
+	sa.syscallsMap = nil
+	sa.bpfModule = nil
+	sa.started = false
+}
+
+func (sa *SeccompAdvisor) update() {
+	if !sa.started {
+		return
+	}
+	for _, c := range sa.containers {
+		v, err := sa.syscallsMap.Lookup(c.Mntns, SyscallsMapValueSize)
+		if err == nil {
+			c.SeccompPolicy = syscallArrToLinuxSeccomp(v)
+		}
+	}
+}
+
+func (sa *SeccompAdvisor) AddContainer(mntns uint64, namespace, podname, containerName string) {
+	sa.containers[mntns] = &types.Container{
+		Mntns:         mntns,
+		Namespace:     namespace,
+		Podname:       podname,
+		ContainerName: containerName,
+	}
+}
+
+func (sa *SeccompAdvisor) RemoveContainer(mntns uint64, namespace, podname, containerName string) {
+	if sa.started {
+		c := sa.containers[mntns]
+		v, err := sa.syscallsMap.Lookup(c.Mntns, SyscallsMapValueSize)
+		if err == nil {
+			c.SeccompPolicy = syscallArrToLinuxSeccomp(v)
+		}
+		err = sa.syscallsMap.Delete(c.Mntns)
```

Comment:
I didn't notice that libbpfgo gained BPFMap functions for lookup and delete: https://github.com/aquasecurity/tracee/pull/422
I should use that instead of my fork.